### ISSUE TITLE
getHashDigest result first character distribution is not uniform

### ISSUE THREAD

**Issue by subzey** [webpack/loader-utils#118]
The string returned from `getHashDigest()` has 0% probabililty of having `0` as the first char.

Other chars from the start of the alphabet: `1`, `2`, ... has a much higher chance to be a first char.

Here's the distribution of the first chars after 10000 trials with random strings ("base64" is used):

character | count
--------- | -----
0 | 0
1 | 25477
2 | 25578
3 | 25132
4 | 420
5 | 389
6 | 382
7 | 402
8 | 406
9 | 393
a | 407
b | 387
c | 383
d | 403
e | 383
f | 389
g | 400
h | 414
i | 404
j | 421
k | 378
l | 394
m | 346
n | 394
o | 390
p | 403
q | 412
r | 395
s | 380
t | 404
u | 416
v | 404
w | 396
x | 388
y | 411
z | 392
A | 415
B | 400
C | 389
D | 410
E | 427
F | 413
G | 374
H | 384
I | 384
J | 401
K | 396
L | 366
M | 419
N | 396
O | 427
P | 385
Q | 377
R | 418
S | 373
T | 409
U | 422
V | 417
W | 382
X | 399
Y | 373
Z | 384
| - | 415
| _ | 372

`getHashDigest` with base64 is often used for generating css modules classnames and it turns out, there's a 75% chance it would start with `1`, `2` or `3`. This is not what you'd expect from a hash, to be honest.

Hashing with other bases has the same issues with the uniformity of the first char.

----

Code for generating the table above:

```javascript
const getHashDigest = require('loader-utils').getHashDigest;
const randomBytes = require('crypto').randomBytes;

const stats = Object.create(null);

for (const letter of '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_') {
	stats[letter] = 0;
}

for (let i = 0; i < 1e5; i++) {
	const digest = getHashDigest(
		randomBytes(64), // 512 bits, enough to expect MD5 128 bit digest to be uniformly random
		'md5', // default
		'base64',
		8
	);

	const firstLetter = digest[0];
	stats[firstLetter]++;
}

console.table(stats);
```

**Comment by subzey** [comment]
TL;DR: [Slice the string](https://github.com/webpack/loader-utils/blob/master/lib/getHashDigest.js#L61-L62) from its tail, not from its head.

----

The long version:

There are two reasons for this.

Each byte of the hash has uniform distribution (the hashing algo is cryptographically strong), but when converted to number and then string, leading zeroes are ignored.

Imagine there's a hash of length 3: `[ 0x01, 0x02, 0x03 ]`. If we convert it into base-64 array, we just take chunks of 6 bits:
`00000001   00000010   00000011` ->
`000000  010000  001000  000011` ("0G83")

and the uniformity of each bit is preserved.

But if we first represent it as a big number, and the as a string, we get "G83". Notice how the leading `0` is gone! This is what happens here, and the hash can never start with a zero.

The way to fix could be preserving leading zeroes during base conversion.

----

The second issue is related to bases that are not powers of two. For example, the hash function is SHA256 and the base is 36. Possible range of hash values is (as hex values):

`ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`
...
`0000000000000000000000000000000000000000000000000000000000000000`

But represented as base-36 it's:

`00000000000000000000000000000000000000000000000000`
...
`6dp5qcb22im238nr3wvp0ic7q99w035jmy2iw7i6n43d37jtof`

As you can see, the hash has no chance to start with `7`...`z`, and the probability of `6` is singificantly lowered.

How to fix this? Ummm... That's not an easy question. From the maths perspective, the best thing can be done is the same as in https://github.com/ai/nanoid: to drop "unfitting" values, get other byte buffer and repeat.

```
/** Pseudocode */
function hash(str, base) {
	const hashLength = Math.floor(256 * Math.LN2 / Math.log(36));
	const maxUniform = BigInt(base) ** BigInt(hashLength) - 1n;

	while (true) {
		const hashAsBigInteger = sha256(str);

		if (hashAsBigInteger <= maxUniform) {
			// We're lucky, value is within uniform range
			return hashAsBigInteger.toString(base).padStart(hashLength, '0');
		}

		// Modify the string and roll the dice once again
		str += '\u0000';
	}
}
```

----

Though, the easiest fix of both issues could be just picking the tail of the string representation:

- We don't care if it was `G83` or `0000000G83`. Jut get the tail and padStart with zeroes.

- In case of SHA256->base36, the first char distribution is highly not uniform. The non-uniformity of last char is something around `1 : 4*10⁷⁵`. That's more than enough for cache busting!
